src/dialogs/sendto/Dialog/FileReceiver.vala

```diff
@@ -0,0 +1,270 @@
+					((Gtk.Window) get_toplevel()).application.send_notification("org.buddiesofbudgie.bluetooth", notification);
+					warning("Error saving transferred file: %s", e.message);
+				}
+				destroy();
+				break;
+		}
+	}
+
+	private void move_to_downloads(string path) throws Error {
+		var source = File.new_for_path(path);
+		var file_name = Path.build_filename(Environment.get_user_special_dir(UserDirectory.DOWNLOAD), source.get_basename());
+		var dest = get_save_name(file_name);
+
+		source.move(dest, FileCopyFlags.ALL_METADATA);
+
+		notification.set_icon(device_image.gicon);
+		notification.set_title(_("File transferred successfully"));
+		notification.set_body(_("Saved file from %s to '%s'").printf(device.alias, dest.get_path()));
+		((Gtk.Window) get_toplevel()).application.send_notification("org.buddiesofbudgie.bluetooth", notification);
+	}
+
+	private File? get_save_name(string uri) {
+		var file = File.new_for_path(uri);
+
+		if (!file.query_exists()) return file;
+
+		var base_name = file.get_basename();
+		var ext_index = base_name.last_index_of(".");
+		var name = ext_index == -1 ? base_name : base_name.substring(0, ext_index);
+		var ext = ext_index == -1 ? "" : base_name.substring(ext_index + 1);
+		var time = new DateTime.now_local().format_iso8601();
+
+		return File.new_for_path(name + " " + time + ext);
+	}
+
+	private void on_transfer_progress(uint64 transferred) {
+		progress_bar.fraction = (double) transferred / (double) total_size;
+		int current_time = (int) get_real_time();
+		int elapsed_time = (current_time - start_time) / 1000000;
+
+		if (current_time < start_time + 1000000) return;
+		if (elapsed_time == 0) return;
+
+		uint64 transfer_rate = transferred / elapsed_time;
+
+		if (transfer_rate == 0) return;
+
+		rate_label.label = Markup.printf_escaped(_("<b>Transfer rate:</b> %s / s"), format_size(transfer_rate));
+		uint64 remaining_time = (total_size - transferred) / transfer_rate;
+		progress_label.label = _("%s of %s received. Time remaining: %s").printf(format_size(transferred), format_size(total_size), format_time((int) remaining_time));
+	}
+
+	private string format_time(int seconds) {
+		if (seconds < 0) seconds = 0;
+		if (seconds < 60) return ngettext("%d second", "%d seconds", seconds).printf(seconds);
+
+		int minutes;
+		if (seconds < 60 * 60) {
+			minutes = (seconds + 30) / 60;
+			return ngettext("%d minute", "%d minutes", minutes).printf(minutes);
+		}
+
+		int hours = seconds / (60 * 60);
+		if (seconds < 60 * 60 * 4) {
+			minutes = (seconds - hours * 60 * 60 + 30) / 60;
+			string h = ngettext("%u hour", "%u hours", hours).printf(hours);
+			string m = ngettext("%u minute", "%u minutes", minutes).printf(minutes);
+			///TRANSLATORS: For example "1 hour, 8 minutes".
+			return _("%s, %s").printf(h, m);
+		}
+
+		return ngettext("about %d hour", "about %d hours", hours).printf(hours);
+	}
```

Comment:
I think it may be worth re-working this function. The initial two conditions are fine, but after that we should probably define minutes as seconds / 60 + a ceil (or a floor, just whether or not we want presentation to be optimistic e.g. report 1 minute when it's technically 1 minute 30s or conservative e.g. report 2 minutes when it might be less, also might play into perceived performance) to drop remainder.

Also curious why we are randomly adding 30s to our seconds.

Once we're at the conditional, we just say if it's minutes < 60, return minutes.

hours + minutes logic seems a bit funky. Am I understanding the conditional after the hours where it's basically if the seconds are less than 4 hours? Any reason for that and not using a modulo to determine total hours then having minutes in the remainder.